pkg/ingester/ingester_v2.go

```diff
@@ -452,3 +461,63 @@ func (i *Ingester) closeAllTSDB() {
 	i.userStatesMtx.Unlock()
 	wg.Wait()
 }
+
+// openExistingTSDB walks the user tsdb dir, and opens a tsdb for each user. This may start a WAL replay, so we limit the number of
+// concurrently opening TSDB.
+func (i *Ingester) openExistingTSDB(ctx context.Context) error {
+	wg := &sync.WaitGroup{}
+	openGate := gate.New(i.cfg.TSDBConfig.MaxOpeningTSDBOnStartup)
+
+	err := filepath.Walk(i.cfg.TSDBConfig.Dir, func(path string, info os.FileInfo, err error) error {
+		if path == i.cfg.TSDBConfig.Dir { // Nothing to do for root
+			return nil
+		}
+
+		// Top level directories are assumed to be user TSDBs
+		if info.IsDir() {
+
+			userID := info.Name()
+			f, err := os.Open(path)
+			if err != nil {
+				level.Error(util.Logger).Log("msg", "unable to open user TSDB dir", "err", err, "user", userID)
+				return filepath.SkipDir
+			}
+			defer f.Close()
+
+			// If the dir is empty skip it
+			if _, err := f.Readdirnames(1); err != nil {
+				if err != io.EOF {
+					level.Error(util.Logger).Log("msg", "unable to read TSDB dir", "err", err, "user", userID)
+					return filepath.SkipDir
+				}
+
+				// Empty dir
+				return filepath.SkipDir
+			}
+
+			// Limit the number of TSDB's opening concurrently
+			if err := openGate.Start(ctx); err != nil {
+				return err
+			}
+
+			wg.Add(1)
```

Comment:
I believe we should add before the call to `Start()`, otherwise you may potentially end up with a race condition where the routines have `Done()` and the gate hasn't started the new one yet (looks an edge case but can be potentially easy to reproduce with a concurrency of `1`).

Reply:
The `filepath.Walk` call won't return till we have crossed this line, i.e. all TSDBs have initiated to open. So, that race condition would not happen I believe.

Reply:
You're right, but actually this comment made me realize that maybe we should move the `openGate.Start()` inside the go routine, so that we can continue iterating the filesystem and scheduling goroutines without being blocked by the gate?

Reply:
I don't see any problem in being blocked by the gate as walking through the files would be fairly cheap compared to replaying the WAL. Additionally, moving it inside the goroutine will make it hard to handle any error caused by `openGate.Start()`. But I am on the fence here.

Reply:
Right. I was primarely thinking loudly.

Reply:
I think I plan to leave as is, because there doesn't seem to be much gain in changing it.